### PR TITLE
Adding regex timeouts, fixing other CA warnings

### LINKED ISSUE #255: Provisioning service preview client missing public API documentation
We cannot enable XML documentation for Provisioning\ServiceClient until all public APIs have been documented.
This is tracking a TODO item within the `Microsoft.Azure.Devices.Provisioning.Service.csproj` file.

### PR DESCRIPTION
**Pull request by CIPop**
## Checklist
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [X] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [X] This pull-request is submitted against the `master` branch.

## Description of the changes
- Adds RegEx timeouts to prevent high CPU in case of malformed input.
- Partial work for #176:
  - Shared is warning-free and treats warnings as errors.
  - \provisioning\* is warning-free and treats warnings as errors.

## Reference/Link to the issue solved with this PR (if any)
Fixes #255

### CODE REVIEW

File: iothub/service/src/IotHubConnectionStringBuilder.cs [hunk 1]
```diff
@@ -17,22 +17,24 @@ namespace Microsoft.Azure.Devices
...
-        static readonly Regex SharedAccessSignatureRegex = new Regex(@"^.+$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
-
-        string hostName;
-        string iotHubName;
-        IAuthenticationMethod authenticationMethod;
+        private const char ValuePairDelimiter = ';';
+        private const char ValuePairSeparator = '=';
+        private const string HostNameSeparator = ".";
+
+        private static readonly string HostNamePropertyName = ((MemberExpression)((Expression<Func<IotHubConnectionStringBuilder, string>>)(_ => _.HostName)).Body).Member.Name; // todo: replace with nameof()
```
Comment (zolvarga) [inline iothub/service/src/IotHubConnectionStringBuilder.cs]: >nameof() [](start = 199, length = 8)

Suggestion: I saw you changed to nameof in other files - maybe that could be done here with the same check in?

File: provisioning/service/src/Config/BulkEnrollmentOperationError.cs [hunk 2]
```diff
@@ -24,14 +24,18 @@ namespace Microsoft.Azure.Devices.Provisioning.Service
...
         /* SRS_BULK_ENROLLMENT_OPERATION_ERRO_21_002: [The BulkEnrollmentOperationError shall store the provided information.] */
 
         /// <summary>
-        /// Registration ID
+        /// Registration ID.
         /// </summary>
+        /// <remarks>
+        /// A valid registration Id shall be alphanumeric, lowercase, and may contain hyphens. Max characters 128.
+        /// </remarks>
+        /// <exception cref="ArgumentException">if the provided string do not fits the registration Id requirements</exception>
```
Comment (zolvarga) [inline provisioning/service/src/Config/BulkEnrollmentOperationError.cs]: >do not fits [](start = 71, length = 11)

Grammar: does not fit

File: provisioning/service/src/Config/EnrollmentGroup.cs [hunk 3]
```diff
@@ -213,8 +213,12 @@ public override string ToString()
         }
 
         /// <summary>
-        /// Enrollment Group ID
+        /// Enrollment Group ID.
         /// </summary>
+        /// <remarks>
+        /// A valid enrollmentGroup Id shall be alphanumeric, lowercase, and may contain hyphens. Max characters 128.
+        /// </remarks>
+        /// <exception cref="ArgumentException">if the provided string do not fits the enrollmentGroup Id requirements</exception>
```
Comment (zolvarga) [inline provisioning/service/src/Config/EnrollmentGroup.cs]: >do not fits  [](start = 71, length = 12)

Grammar...

File: provisioning/service/src/Config/IndividualEnrollment.cs [hunk 4]
```diff
@@ -191,8 +191,12 @@ public override string ToString()
         }
 
         /// <summary>
-        /// Registration ID
+        /// Registration ID.
         /// </summary>
+        /// <remarks>
+        /// A valid registration Id shall be alphanumeric, lowercase, and may contain hyphens. Max characters 128.
+        /// </remarks>
+        /// <exception cref="ArgumentException">if the provided string do not fits the registration Id requirements</exception>
```
Comment (zolvarga) [inline provisioning/service/src/Config/IndividualEnrollment.cs]: >do not fits [](start = 71, length = 11)

Grammar...

File: provisioning/service/src/Config/BulkEnrollmentOperationError.cs (hunk 2 above)
Comment (CIPop) [reply provisioning/service/src/Config/BulkEnrollmentOperationError.cs]: Copy paste...

File: iothub/service/src/IotHubConnectionStringBuilder.cs (hunk 1 above)
Comment (CIPop) [reply iothub/service/src/IotHubConnectionStringBuilder.cs]: Actually I don't agree with the original comment. I don't think there is any connection between the Azure ConnectionString format and the API names. They should actually be simple strings like "HostName", etc.

**Pull request merged**